workflow/docs/design.md

```diff
@@ -0,0 +1,406 @@
+> - While an "Actor" failed, the whole workflow instance is trapped into failed.
+
+The Actor will do all "dirty" effects, for example: creating Chaos Resource, running user-defined tasks, and so on.
+
+```go
+type Actor interface {
+  PlayOn(pg Playground) error
+}
+
+type Playground interface {
+  CreateNetworkChaos(networkChaos chaosmeshv1alph1.NetworkChaos) error
+  DeleteNetworkChaos(namespace, name string) error
+  CreatePodChaos(podChaos chaosmeshv1alph1.PodChaos) error
+  DeletePodChaos(namespace, name string) error
+
+  // Other ad-hoc methods
+}
+```
+
+### WorkflowManager: rest glue codes
+
+We need a manager who makes these components work together, and controls the main-loop, and update the status of the workflow, and max-concurrency, and so-on.
+
+The main loop looks like:
+
+```go
+func loop(){
+  for {
+    workflowKey, err := trigger.AcquireNext()
+    scheduler, err := fetchSchedulerFor(workflowKey)
+
+    go func(){
+      templates, err := scheduler.scheduleNext()
+      for item := range templates {
+        go func (){
+          // TODO: update node status
+          actor := fetchActor(item)
+          err := actor.Apply()
+          // TODO: update node status
+          if item.NeedReQueue() {
+            // TODO: manually ask for next trigger
+          }
+        }()
+      }
+    }
+  }
+}
+```
+
+### States for Node
+
+There are 8 available phase of one Node:
+
+- Init
+- WaitingForSchedule
+- WaitingForChild
+- Running
+- Evaluating
+- Holding
+- Succeed
+- Failed
+
+**Init** means is the default phase when **Node** just created, means this node is just created, did not effect real world yet.
+
+**WaitingForSchedule** is only available on Serial, Parallel and Task, it means this Node is idle and safe for next scheduling; It is presents:
+
+- For Chaos, Suspend, Task, Parallel:
+  - This Node did not make "Effect" yet.
+- For Serial:
+  - This Node did not create child node yet;
+  - Or previous child node just succeed;
+
+**WaitingForChild** is only available on Serial, Parallel and Task, it means at least 1 child node is in **Running** state.
+
+**Running** is available on type Chaos, Suspend, Task; It means an **Actor** is doing dirty work for this node. For Chaos, both of "create Chaos CRD resource" and "delete Chaos CRD resource" are presented as **Running**.
+
+**Evaluating** is only available on Task, it means Task is collecting result of user's pod, then picks templates to execute.
+
+> Question: Should we split it?
+
+**Holding** is available on Chaos, Suspend; It means current node is waiting for next action. For example: a Chaos node which in **Holding** is waiting for the end of this ChaosExperiments, then delete it.
+
+**Succeed** means this node is completed.
+
+**Failed** means this node failed. One **Failed** node will cause the whole Workflow fall into **Failed**.
+
+Examples:
+
+A **NetworkChaos** Node: **Init** -> **Running** -> **Holding** -> **Running** -> **Succeed**
```

Comment:
What will happen if the user delete the created `NetworkChaos` manually? Will it break the status machine?

Reply:
I did not think of that before. 🤔 

Because of the pattern of kubernetes controller, I think workflow SHOULD re-create a new `NetworkChaos` CRD.

I will append for this in future PRs.